integration_test/trackedusersreporting/tracked_users_reporting_test.go

```diff
@@ -0,0 +1,298 @@
+package trackedusersreporting
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	nethttptest "net/http/httptest"
+	"os"
+	"path"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/testhelper/transformertest"
+
+	"github.com/segmentio/go-hll"
+
+	"github.com/ory/dockertest/v3"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+	kithttputil "github.com/rudderlabs/rudder-go-kit/httputil"
+	kithelper "github.com/rudderlabs/rudder-go-kit/testhelper"
+	"github.com/rudderlabs/rudder-go-kit/testhelper/docker/resource/postgres"
+	"github.com/rudderlabs/rudder-server/runner"
+	"github.com/rudderlabs/rudder-server/testhelper/backendconfigtest"
+	"github.com/rudderlabs/rudder-server/testhelper/health"
+	webhookutil "github.com/rudderlabs/rudder-server/testhelper/webhook"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	writeKey    = "writekey-1"
+	sourceID    = "source-1"
+	workspaceID = "workspace-1"
+)
+
+type testConfig struct {
+	postgresResource *postgres.Resource
+	gwPort           int
+	webhook          *webhookutil.Recorder
+	configBEServer   *nethttptest.Server
+	transformerUrl   string
+}
+
+type userIdentifier struct {
+	userID      string
+	anonymousID string
+}
+
+func TestTrackedUsersReporting(t *testing.T) {
+	tc := setup(t)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	wg, ctx := errgroup.WithContext(ctx)
+	wg.Go(func() error {
+		err := runRudderServer(t, ctx, tc.gwPort, tc.postgresResource, tc.configBEServer.URL, t.TempDir(), tc.transformerUrl)
+		if err != nil {
+			t.Logf("rudder-server exited with error: %v", err)
+		}
+		return err
+	})
+
+	url := fmt.Sprintf("http://localhost:%d", tc.gwPort)
+	health.WaitUntilReady(ctx, t, url+"/health", 60*time.Second, 10*time.Millisecond, t.Name())
+
+	eventsCount, err := sendEvents([]userIdentifier{
+		{userID: "user-1", anonymousID: "anon-1"},
+		{userID: "user-2", anonymousID: "anon-2"},
+		{userID: "user-1"},
+		{anonymousID: "anon-1"},
+		{userID: "user-2"},
+		{anonymousID: "anon-2"},
+	}, "identify", writeKey, url)
+	require.NoError(t, err)
+
+	require.Eventuallyf(t, func() bool {
+		return tc.webhook.RequestsCount() == eventsCount
+	}, 1*time.Minute, 5*time.Second, "unexpected number of events received, count of events: %d", tc.webhook.RequestsCount())
+
+	// TODO: once flusher is implemented, add a mock reporting server to check is we are getting correct cardinality of users
+	cardinalityMap := getCardinalityFromDB(t, tc.postgresResource)
+	require.Equal(t, 2, cardinalityMap[workspaceID][sourceID])
+	cancel()
+	require.NoError(t, wg.Wait())
+}
+
+func getCardinalityFromDB(t *testing.T, postgresResource *postgres.Resource) map[string]map[string]int {
```

Comment:
This might stop working when flusher code gets introduced. Given that fluster will get and then delete them from DB. So we will need to adapt the test to mock the reporting server instead of reading the DB.